File: OpenBurnBarCore/Sources/OpenBurnBarComputerUseCore/AppCheckAttestationBinding.swift
import CryptoKit
import Foundation

/// Canonical App Check attestation digest shared by Mac, iOS, and Cloud Functions (WS2/WS4).
/// Wire field `attestationHashBlake3` carries this SHA-256 hex (name is historical).
public enum AppCheckAttestationBinding {
    public static let claimKey = "obb_app_check"
    public static let canonicalPrefix = "openburnbar.appcheck.v1"
    public static let maxAgeMillis: Int64 = 30 * 24 * 60 * 60 * 1000

    public struct Claim: Equatable, Sendable {
        public let appId: String
        public let boundAtMillis: Int64

        public init(appId: String, boundAtMillis: Int64) {
            self.appId = appId
            self.boundAtMillis = boundAtMillis
        }
    }

    public static func digestHex(appId: String, boundAtMillis: Int64) -> String {
        let payload = "\(canonicalPrefix)|\(appId)|\(boundAtMillis)"
        let hash = SHA256.hash(data: Data(payload.utf8))
        return hash.map { String(format: "%02x", $0) }.joined()
    }

    public static func digestHex(for claim: Claim) -> String {
        digestHex(appId: claim.appId, boundAtMillis: claim.boundAtMillis)
    }

    public static func parseClaim(from token: [String: Any]?) -> Claim? {
        guard let raw = token?[claimKey] as? [String: Any],
              let appId = raw["appId"] as? String,
              !appId.isEmpty,
              let boundAtMillis = raw["boundAtMillis"] as? Int64
              ?? (raw["boundAtMillis"] as? Double).map({ Int64($0) }) else {
            return nil
        }
        return Claim(appId: appId, boundAtMillis: boundAtMillis)
    }

    public static func isFresh(_ claim: Claim, nowMillis: Int64 = Int64(Date().timeIntervalSince1970 * 1000)) -> Bool {
        nowMillis - claim.boundAtMillis <= maxAgeMillis
    }
}

File: android/app/src/main/java/com/openburnbar/data/computeruse/AppCheckAttestationBinding.kt
package com.openburnbar.data.computeruse

import java.security.MessageDigest

/**
 * Canonical App Check attestation digest shared by Mac, iOS, Cloud Functions,
 * and Android (WS2/WS4). 1:1 Kotlin port of the Swift
 * `AppCheckAttestationBinding`
 * (`OpenBurnBarCore/Sources/OpenBurnBarComputerUseCore/AppCheckAttestationBinding.swift`).
 *
 * The server's `bindAppCheckAttestation` callable verifies the caller's App
 * Check token (Play Integrity on Android, App Attest on iOS) and writes the
 * platform-agnostic `obb_app_check` custom claim (`appId`, `boundAtMillis`)
 * onto the Firebase Auth user. Clients re-derive the SHA-256 hex digest from
 * that claim and carry it as the wire field `attestationHashBlake3` (the
 * **controller app's** digest — the field name is historical). Mac strict
 * mode checks presence, not equality with the Mac host App Check app id.
 */
object AppCheckAttestationBinding {
    const val CLAIM_KEY = "obb_app_check"
    const val CANONICAL_PREFIX = "openburnbar.appcheck.v1"
    const val MAX_AGE_MILLIS: Long = 30L * 24L * 60L * 60L * 1000L

    data class Claim(
        val appId: String,
        val boundAtMillis: Long,
    )

    fun digestHex(appId: String, boundAtMillis: Long): String {
        val payload = "$CANONICAL_PREFIX|$appId|$boundAtMillis".toByteArray(Charsets.UTF_8)
        return MessageDigest.getInstance("SHA-256").digest(payload)
            .joinToString("") { "%02x".format(it.toInt() and BYTE_MASK) }
    }

    fun digestHex(claim: Claim): String = digestHex(appId = claim.appId, boundAtMillis = claim.boundAtMillis)

    /**
     * Parse the binding claim from ID-token claims. Mirrors the Swift parser:
     * absent/blank `appId` or a non-numeric `boundAtMillis` ⇒ null (no claim).
     */
    fun parseClaim(claims: Map<String, Any?>?): Claim? {
        val raw = claims?.get(CLAIM_KEY) as? Map<*, *> ?: return null
        val appId = (raw["appId"] as? String)?.takeIf { it.isNotEmpty() } ?: return null
        val boundAtMillis = (raw["boundAtMillis"] as? Number)?.toLong() ?: return null
        return Claim(appId = appId, boundAtMillis = boundAtMillis)
    }

    fun isFresh(claim: Claim, nowMillis: Long = System.currentTimeMillis()): Boolean = nowMillis - claim.boundAtMillis <= MAX_AGE_MILLIS

    private const val BYTE_MASK = 0xFF
}
